xkcd/pkg/index/index.go

```diff
@@ -0,0 +1,158 @@
+package index
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/vespian/go-exercises/xkcd/pkg/pbuff"
+	"github.com/vespian/go-exercises/xkcd/pkg/types"
+	"github.com/vespian/go-exercises/xkcd/pkg/web"
+)
+
+func serialize(a types.AllStories, t types.OndiskSerialization,
+) (
+	blob []byte,
+	err error,
+) {
+	switch t {
+	case types.JSON:
+		blob, err = json.Marshal(a)
+		if err != nil {
+			return nil, fmt.Errorf("Index JSON marshaling failed: %s", err)
+		}
+	case types.Protobuf:
+		pbuffDigestableStructs := pbuff.PBAllStoriesFromAllStories(a)
+		blob, err = proto.Marshal(pbuffDigestableStructs)
+		if err != nil {
+			log.Fatal("pbuff marshaling error: ", err)
+		}
+	default:
+		panic("Unsupported serializing method")
+	}
+
+	return
+}
+
+func deserialize(in []byte, t types.OndiskSerialization,
+) (
+	res types.AllStories,
+	err error,
+) {
+	switch t {
+	case types.JSON:
+		if err = json.Unmarshal(in, &res); err != nil {
+			log.Fatalf("JSON unmarshaling failed: %s", err)
+		}
+	case types.Protobuf:
+		tmp := new(pbuff.PBAllStories)
+
+		err = proto.Unmarshal(in, tmp)
+		if err != nil {
+			log.Fatal("pbuff unmarshaling error: ", err)
+		}
+		res = pbuff.AllStoriesFromPBAllStories(tmp)
+	default:
+		panic("Unsupported serializing method")
+	}
+
+	return
+}
+
+func read(idx types.IndexFile) (types.AllStories, error) {
+	var blob []byte
+	var a types.AllStories
+	var err error
+
+	if blob, err = ioutil.ReadFile(idx.Location); err != nil {
+		return nil, err
+	}
+
+	if a, err = deserialize(blob, idx.Type); err != nil {
+		return nil, err
+	}
+
+	return a, nil
+}
+
+func store(a types.AllStories, idx types.IndexFile) (err error) {
```

Comment:
this is just a style thing, and there's no right way, but here's how i would do this anyway...

```
func store(stories types.AllStories, idx types.IndexFile) error {
  buf, err := serialize(stories, idx.Type)
  if err != nil {
    return err
  }
  return ioutil.WriteFile(idx.Location, buf, os.FileMode(0540))
}
```